californium-core/src/test/java/org/eclipse/californium/rule/CoapNetworkRule.java

```diff
@@ -0,0 +1,144 @@
+	private static final int DEFAULT_MESSAGE_THREADS = 1;
+	/**
+	 * CoAP datagram formatter. Used for logging.
+	 */
+	private static final DatagramFormatter FORMATTER = new DatagramFormatter() {
+
+		private DataParser parser = new UdpDataParser();
+
+		@Override
+		public String format(byte[] data) {
+			Message message = parser.parseMessage(data);
+			return message.toString();
+		}
+
+	};
+
+	/**
+	 * Number of message processing threads. Used to setup test
+	 * {@link NetworkConfig}.
+	 * 
+	 * @see #createTestConfig()
+	 * @see #createStandardTestConfig()
+	 * @see #getStandardTestConfig()
+	 */
+	private final AtomicInteger messageThreads = new AtomicInteger(DEFAULT_MESSAGE_THREADS);
+
+	/**
+	 * Create rule supporting provided modes.
+	 * 
+	 * @param modes supported datagram socket implementation modes.
+	 */
+	public CoapNetworkRule(Mode... modes) {
+		super(FORMATTER, modes);
+		this.messageThreads.set(DEFAULT_MESSAGE_THREADS);
+	}
+
+	/**
+	 * Set number of message processing threads. Using multiple threads for
+	 * sending and receiving may result in reordering of messages. Therefore the
+	 * default is {@link #DEFAULT_MESSAGE_THREADS} to ensure, that no such
+	 * reorder happens.
+	 * 
+	 * @param threads number of threads.
+	 * @return this rule
+	 */
+	public CoapNetworkRule setMessageThreads(int threads) {
+		if (1 > threads) {
+			throw new IllegalArgumentException("number of message threads must be at least 1, not " + threads + "!");
+		}
+		messageThreads.set(threads);
+		return this;
+	}
+
+	@Override
+	protected void initNetwork(boolean first) {
+		if (first) {
+			EndpointManager.reset();
+		}
+		createStandardTestConfig();
+		super.initNetwork(first);
+	}
+
+	@Override
+	protected void closeNetwork() {
+		super.closeNetwork();
+		EndpointManager.reset();
+		messageThreads.set(DEFAULT_MESSAGE_THREADS);
+		NetworkConfig.setStandard(null);
+	}
+
+	/**
+	 * Create new standard network configuration for testing.
+	 * 
+	 * @return fresh standard network configurations. Changes are visible to
+	 *         other usage of the standard configuration.
+	 * @see NetworkConfig#getStandard()
+	 */
+	public NetworkConfig createStandardTestConfig() {
+		NetworkConfig config = createTestConfig();
+		NetworkConfig.setStandard(config);
+		return config;
+	}
+
+	/**
+	 * Get standard network configuration for testing.
+	 * 
+	 * @return standard network configurations. Changes are visible to other
+	 *         usage of the standard configuration.
+	 * @see NetworkConfig#getStandard()
+	 */
+	public NetworkConfig getStandardTestConfig() {
+		return NetworkConfig.getStandard();
+	}
+
+	/**
+	 * Create new network configuration for testing.
+	 * 
+	 * @return network configurations. Detached from the standard network
+	 *         configuration.
+	 */
+	public NetworkConfig createTestConfig() {
+		int threads = messageThreads.get();
+		NetworkConfig config = new NetworkConfig();
```

Comment:
Looking on the test, I was not sure, if all test are using the `NetworkConfig` in a "proper intended way" or more "in random way". 
Therefore my intention is to move the access of `NetworkConfig` to `CoapNetworkRule` to put that configuration under the "scope control" of the rule and to make it easier, the search for "out of scope usage". 
So three methods are introduced with (in my opinion) proper documentation about how to use it and explicitly mentioning the side effect ("Changes are visible to other usage of the standard configuration").
So, if currently `createTestConfig` is not used, this is not really "by design", it's more just as its used now. I just want to make sure, that nobody uses `NetworkConfig` directly.